crates/blockchain/blockchain.rs

```diff
@@ -1892,56 +1917,46 @@ pub fn get_total_blob_gas(tx: &EIP4844Transaction) -> u32 {
     GAS_PER_BLOB * tx.blob_versioned_hashes.len() as u32
 }
 
-/// Collapses a root branch node into an extension or leaf node if it has only one valid child.
-/// Returns None if there are no valid children.
-///
-/// NOTE: this assumes the branch has 0 or 1 children. If there are more than 1,
-/// it will discard all but the first valid child found.
-#[cold]
-fn collapse_root_node(
-    real_root: Box<BranchNode>,
-    state_updates_map: &mut FxHashMap<Nibbles, Vec<u8>>,
-    storage: &Store,
-    parent_header: &BlockHeader,
-) -> Result<Option<Node>, StoreError> {
-    // Collapse the branch into an extension or leaf
-    let Some((choice, only_child)) = real_root
-        .choices
-        .into_iter()
-        .enumerate()
-        .find(|(_, c)| c.is_valid())
-    else {
-        return Ok(None);
-    };
-    let path = Nibbles::from_hex(vec![choice as u8]);
-    let child_bytes = match state_updates_map.get(&path) {
-        Some(v) => v.clone(),
-        None => storage
-            .state_trie(parent_header.hash())?
-            .ok_or(StoreError::MissingStore)?
-            .db()
-            .get(path)?
-            .ok_or_else(|| StoreError::Custom("Missing child node during root collapse".into()))?,
-    };
-    // Same match as in [`BranchNode::remove`]
-    let child = match Node::decode(&child_bytes)? {
-        // Replace root with an extension node leading to the child
-        Node::Branch(_) => {
-            ExtensionNode::new(Nibbles::from_hex(vec![choice as u8]), only_child).into()
-        }
-        // Replace root with the child extension node, updating its path in the process
-        Node::Extension(mut extension_node) => {
-            let mut extension_node = extension_node.take();
-            extension_node.prefix.prepend(choice as u8);
-            extension_node.into()
+fn branchify(node: Node) -> BranchNode {
```

Comment:
It makes sense to keep the `BranchNode` in a box, I think:

```suggestion
/// Turns [`Node`] into a [`BranchNode`], if needed, and returns it
fn branchify(node: Node) -> Box<BranchNode> {
```

The only use of this function is in `collect_trie`, and after calling this function it puts the result in a `Box`